tests/main/remodel-gadget/task.yaml

```diff
@@ -0,0 +1,124 @@
+environment:
+    OLD_GADGET: pc
+    NEW_GADGET: test-snapd-pc
+
+systems: [ubuntu-core-18-64]
+
+prepare: |
+    if [ "$TRUST_TEST_KEYS" = "false" ]; then
+        echo "This test needs test keys to be trusted"
+        exit
+    fi
+    #shellcheck source=tests/lib/systemd.sh
+    . "$TESTSLIB"/systemd.sh
+    systemctl stop snapd.service snapd.socket
+    rm -rf /var/lib/snapd/assertions/*
+    rm -rf /var/lib/snapd/device
+    rm -rf /var/lib/snapd/state.json
+    mv /var/lib/snapd/seed/assertions/model model.bak
+    cp "$TESTSLIB"/assertions/developer1.account /var/lib/snapd/seed/assertions
+    cp "$TESTSLIB"/assertions/developer1.account-key /var/lib/snapd/seed/assertions
+    cp "$TESTSLIB"/assertions/developer1-pc-18.model /var/lib/snapd/seed/assertions
+    cp "$TESTSLIB"/assertions/testrootorg-store.account-key /var/lib/snapd/seed/assertions
+    # kick first boot again
+    systemctl start snapd.service snapd.socket
+    while ! snap changes | grep -q "Done.*Initialize system state"; do sleep 1; done
+
+restore: |
+    if [ "$TRUST_TEST_KEYS" = "false" ]; then
+        echo "This test needs test keys to be trusted"
+        exit
+    fi
+    #shellcheck source=tests/lib/systemd.sh
+    . "$TESTSLIB"/systemd.sh
+    systemctl stop snapd.service snapd.socket
+    rm -rf /var/lib/snapd/assertions/*
+    rm -rf /var/lib/snapd/device
+    rm -rf /var/lib/snapd/state.json
+    rm -f /var/lib/snapd/seed/assertions/developer1.account
+    rm -f /var/lib/snapd/seed/assertions/developer1.account-key
+    rm -f /var/lib/snapd/seed/assertions/developer1-pc-18.model
+    rm -f /var/lib/snapd/seed/assertions/testrootorg-store.account-key
+    cp model.bak /var/lib/snapd/seed/assertions/model
+    rm -f ./*.bak
+    # kick first boot again
+    systemctl start snapd.service snapd.socket
+    # wait for first boot to be done
+    snap wait system seed.loaded
+    # cleanup the canary file
+    rm -f /boot/efi/canary.txt
+    while ! snap changes | grep -q "Done.*Initialize system state"; do sleep 1; done
+    # extra paranoia because failure to cleanup earlier took us a long time
+    # to find
+    if [ -e /var/snap/$NEW_GADGET/current ]; then
+        echo "Leftover $NEW_GADGET data dir found, test does not "
+        echo "properly cleanup"
+        echo "see https://github.com/snapcore/snapd/pull/6620"
+        echo
+        find /var/snap
+        exit 1
+    fi
+execute: |
+    if [ "$TRUST_TEST_KEYS" = "false" ]; then
+        echo "This test needs test keys to be trusted"
+        exit
+    fi
+    #shellcheck source=tests/lib/boot.sh
+    . "$TESTSLIB"/boot.sh
+    wait_change_done() {
+        chg_summary="$1"
+        for _ in $(seq 10); do
+            if snap changes | grep -qE "[0-9]+\\ +Done\\ +.* $chg_summary"; then
+                break
+            fi
+            # some debug output
+            snap changes
+            # wait a bit
+            sleep 5
+        done
+        snap changes | MATCH "$chg_summary"
```

Comment:
This check is less precise than the grep match inside the retry loop; shouldn't we check the same pattern here to ensure the change is "done"?